### PR TITLE
 Exchange & Proxy Integration Tests - ERC1155

### PR DESCRIPTION
**Pull request by hysz**
## Description

Integration tests for the ERC1155 proxy with the Mutli-Asset Proxy and Exchange.

## Checklist:

-   [ ] Prefix PR title with `[WIP]` if necessary.
-   [ ] Add tests to cover changes as needed.
-   [ ] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/asset-proxy/contracts/src/ERC1155Proxy.sol [hunk 1]
```diff
@@ -0,0 +1,268 @@
+/*
```
Comment (fabioberger) [inline contracts/asset-proxy/contracts/src/ERC1155Proxy.sol]: How this already been reviewed elsewhere? Are we duplicating Solidity source code within the monorepo?

File: packages/types/CHANGELOG.json [hunk 2]
```diff
@@ -1,4 +1,22 @@
...
+    },
+    {
+        "version": "2.2.2",
+        "changes": [
+            {
+                "note": "Added ERC1155 revert reasons",
+                "pr": 1657
+            }
+        ]
+    },
```
Comment (fabioberger) [inline packages/types/CHANGELOG.json]: These should be one CHANGELOG entry.

**Comment by fabioberger** [comment]
Eventually we'll also need to add ERC1155 support to OrderWatcher.

File: contracts/asset-proxy/contracts/src/ERC1155Proxy.sol (hunk 1 above)
Comment (hysz) [reply contracts/asset-proxy/contracts/src/ERC1155Proxy.sol]: Yup! Been reviewed in #1661. Just had to do a cheeky rebase.

File: contracts/exchange/test/core.ts [hunk 3]
```diff
@@ -1040,6 +1086,435 @@ describe('Exchange core', () => {
...
+                // takerAddress / takerToken[1]
+                expectedInitialBalances[10].minus(totalTakerValuesTransferred[1]),
+                // takerAddress / takerToken[2]
+                expectedInitialBalances[11].minus(totalTakerValuesTransferred[2]),
+            ];
+            await erc1155Wrapper.assertBalancesAsync(tokenHolders, tokensToTransfer, expectedFinalBalances);
+        });
+        it('should allow an order exchanging erc1155 assets to be partially filled', async () => {
+            // NOTICE:
+            // As-per the eip1155 standard, there is no way to distinguish between a fungible or non-fungible erc1155 assets.
```
Comment (abandeali1) [inline contracts/exchange/test/core.ts]: Shouldn't a partial transfer of a non-fungible still fail with `TRANSFER_FAILED`? Whatever the behavior is, we should add a test for it.

Comment (hysz) [reply contracts/exchange/test/core.ts]: We chatted about this offline.

Partial fills on non-fungible assets should be prohibited by the maker by setting the amount to 1. The proxy cannot distinguish between fungible and non-fungible assets, so it cannot force partial fills to fail in this scenario. If the amount is _not_ set to 1 and an asset is non-fungible then the behavior depends on the implementation of the ERC1155 contract.

**Pull request merged**